lib/monitoring/middleware/conjur_exporter.rb

```diff
@@ -0,0 +1,93 @@
+require 'prometheus/client'
+require 'prometheus/client/formats/text'
+
+module Monitoring
+  module Middleware
+    # Exporter is a Rack middleware that provides a sample implementation of a
+    # Prometheus HTTP exposition endpoint.
+    #
+    # By default it will export the state of the global registry and expose it
+    # under `/metrics`. Use the `:registry` and `:path` options to change the
+    # defaults.
+    class ConjurExporter
+      attr_reader :app, :path
+
+      FORMATS  = [Prometheus::Client::Formats::Text].freeze
+      FALLBACK = Prometheus::Client::Formats::Text
+
+      def initialize(app, options = {})
+        @app = app
+        @path = options[:path] || '/metrics'
+        @port = options[:port]
+        @acceptable = build_dictionary(FORMATS, FALLBACK)
+      end
+
+      def call(env)
+        if metrics_port?(env['SERVER_PORT']) && env['PATH_INFO'] == @path
+          format = negotiate(env, @acceptable)
+          format ? respond_with(format) : not_acceptable(FORMATS)
+        else
+          @app.call(env)
+        end
+      end
+
+      private
+
+      def negotiate(env, formats)
+        parse(env.fetch('HTTP_ACCEPT', '*/*')).each do |content_type, _|
+          return formats[content_type] if formats.key?(content_type)
+        end
+
+        nil
+      end
+
+      def parse(header)
+        header.split(/\s*,\s*/).map do |type|
+          attributes = type.split(/\s*;\s*/)
+          quality = extract_quality(attributes)
+
+          [attributes.join('; '), quality]
+        end.sort_by(&:last).reverse
+      end
+
+      def extract_quality(attributes, default = 1.0)
+        quality = default
+
+        attributes.delete_if do |attr|
+          quality = attr.split('q=').last.to_f if attr.start_with?('q=')
+        end
+
+        quality
+      end
+
+      def respond_with(format)
+        [
+          200,
+          { 'Content-Type' => format::CONTENT_TYPE },
+          [format.marshal(Monitoring::Metrics.registry)]
```

Comment:
Thoughts about dependency injecting this via the constructor. I think it makes this easier to test and decouples it from the Metrics module. It looks like other parts of the code expect this to be passed in as an option

Reply:
Addressed in #[0417de0](https://github.com/cyberark/conjur/pull/2501/commits/0417de010938d472b8e2d16d4c142bb09e71928a)